# Tipos de Dado

## O que é um Dado?

Dado é tudo aquilo que guarda informação, ou seja, um numero, um texto, Verdadeiro ou Falso, uma letra, qualquer coisa que seja informação

Eles são divididos entre Primitivos (ou Simples, ou Básicos) e Compostos (ou Agrupados, ou Complexos)

## Dados Primitivos

Dados Primitivos guardam apenas uma informação, ou seja, um numero, um texto, um Verdadeiro

Eles são divididos em 3 tipos:

- Strings (str)
- Integers e Floats (int & float)
- Booleans (bool)

### String (Textos)

#### Declaração

As Strings representam os textos, sua declaração é feita através de aspas duplas ("") ou unicas ('')

In [5]:
nome = "Mirai"
linguagem = 'Python'
print(nome, linguagem)

Mirai Python


Nós não podemos misturar as aspas, isto é, começar uma declaração com aspas duplas ("") e terminar com aspas unicas ('')

nós podemos dentro da string usar a aspa que não foi usada durante a declaração, isto se torna util em texto em Ingles

In [4]:
apresentacao = "Hi, my name is Mirai, I'm a Developer" # Note o uso das aspas unicas dentro da string
citacao = 'Pois então ela disse, "Não se dirija mais a mim"...' # Note o uso das aspas duplas dentro da string

# repare que envolta do texto usamos o mesmo tipo de aspas
print(apresentacao, citacao)

Hi, my name is Mirai, I'm a Developer Pois então ela disse, "Não se dirija mais a mim"...


porém, nos podemos sim usar o mesmo tipo de aspas dentro e fora da string, basta ultilizarmos o caractere de escapagem (\\)

In [3]:
apresentacao = 'Hi, my name is Mirai, I\'m a Developer'
citacao = "Pois então ela disse, \"Não se dirija mais a mim\"..."
print(apresentacao, citacao)

Hi, my name is Mirai, I'm a Developer Pois então ela disse, "Não se dirija mais a mim"...


#### Multi-Line

Tambem temos as "Multi-line Strings", elas são declaradas através de 3 aspas, sejam elas unicas ou duplas, elas são textos que se espalham por multiplas linhas, preservando espaços

In [2]:
multi_line1 = """Olá, eu sou um texto de multiplas linhasl, eu fui declarado usando aspas duplas \"\"

  eu preservo os espaços

    eba
"""

multi_line2 = '''Olá, eu sou um texto de multiplas linhas, eu fui declarado usando aspas unicas \'\'

  eu preservo os espaços

    eba
'''

print(multi_line1)
print(multi_line2)

Olá, eu sou um texto de multiplas linhasl, eu fui declarado usando aspas duplas ""

  eu preservo os espaços

    eba

Olá, eu sou um texto de multiplas linhas, eu fui declarado usando aspas unicas ''

  eu preservo os espaços

    eba



#### fStrings (Formatted Strings)

as fStrings, são strings formatadas, isto é, strings que contém codigo, sua declaração é feita adicionando um `f` a frente das aspas, e envolvendo os codigos em chaves `{}`

In [7]:
nome = "Mirai"
linguagem = "Python"
idade = 17
apresentacao = f"Olá, meu nome é {nome}, eu tenho {idade} anos e sei programar em {linguagem}"
print(apresentacao)

Olá, meu nome é Mirai, eu tenho 17 anos e sei programar em Python


eu disse codigo, pois podemos adicionar coisas além de variaveis, como funções, operações aritmeticas, metodos, etc...

Eis aqui um exemplo:

In [8]:
nome = "Mirai"
linguagem = "Python"
idade = 17
apresentacao = f'Olá, meu nome é {nome.upper()}, eu nasci em {2022 - idade} e sei programar em {linguagem}'
print(apresentacao)

Olá, meu nome é MIRAI, eu nasci em 2005 e sei programar em Python


As fStrings, podem ser misturadas com as Multi-Line strings, basta adicionar um `f` a frente

elas tambem podem ser feitas através do string.format(), este metodo pega os argumentos providos e substutui nos espaços reservados em chaves `{}`

In [9]:
apresentacao = 'Olá, meu nome é {}, eu nasci em {} e sei programar em {}'
nome = "Mirai"
linguagem = "Python"
idade = 17
print(apresentacao.format(nome, idade, linguagem))

Olá, meu nome é Mirai, eu nasci em 17 e sei programar em Python


#### Concatenação

Concatenação é o ato de juntar duas Strings, veja o mesmo exemplo anterior porém concatenado

In [10]:
nome = "Mirai"
linguagem = "Python"
idade = 17
apresentacao = "Olá, meu nome é " + nome + ", eu tenho " + str(idade) + " anos e sei programar em " + linguagem
print(apresentacao)

Olá, meu nome é Mirai, eu tenho 17 anos e sei programar em Python


Note que nós devemos fazer a devida conversão para o tipo string (str), e devemos manejar os espaços manualmente, o que pode ser deveras entediante

### Numeros (Ints & Floats) e Operações Aritméticas

#### Ints

INTs, Integers ou Integros, são quaisquer numeros inteiros, eles são passiveis de operações aritmeticas

In [13]:
idade = 17
ano_de_nascimento = 2005
ano_atual = 2022
print(type(idade))

<class 'int'>


#### Floats

FLOATs, Floating Point Numbers ou Numeros de Pontos Flutuantes, são quaisquer numeros com casas decimais, ou um ponto (`.`) em sua declaração

In [14]:
PI = 3.1415
print(type(PI))

<class 'float'>


#### Operações Aritméticas

Em Python se é possivel exercer operações aritméticas, eis aqui uma lista das possiveis operações

- \+ -> soma
- \- -> subtração
- ** -> potencia
- \* -> mutiplicação
- / -> divisão
- // -> "divisão verdadeira"
- % -> modulo

Python segue sua própria ordem de precedencia baseada nos metodos BODMAS e PEMDAS, eis aqui sua ordem:

- 1 - PARENTESES ((())) -> DE DENTRO PRA FORA
- 2 - POTENCIA
- 3 - RAIZES
- 4 - MUTIPLICAÇÃO
- 5 - DIVISÃO
- 6 - SOMA
- 7 - SUBTRAÇÃO

As operações aritméticas são as mesmas aprendidas na escola, porém irei ressaltar 3, divisão, "divisão verdadeira" e modulo

##### Divisão

Em Python uma Divisão sempre resultará em um float, mesmo que o resultado seja inteiro:

In [17]:
print(4/2, type(4/2))
print(1/3, type(1/3))

2.0 <class 'float'>
0.3333333333333333 <class 'float'>


Isso pode te levar a problemas caso necessite de um numero de tipo `int`, para isso se existe a "Divisão Verdadeira"

##### "Divisão Verdadeira"

A "divisão verdadeira" também é conhecida como "divisão arredondada" ou "divisão exclusiva", é uma operação de divisão que converte o resultado para Int

o que ela essencialmente faz é cortar os numeros após o ponto, "arredondando" assim o numero

In [2]:
print(4//2, type(4//2))
print(1//3, type(1//3))

2 <class 'int'>
0 <class 'int'>


##### Problemas com Numeros de Ponto Flutuante

##### Bit Overflow

Qual o resultado da soma `0.1 + 0.2`?

Caso você performe a soma em uma calculadora o resultado será `0.3`, mas observe esta mesma operação em Python

In [1]:
print(0.1 + 0.2)

0.30000000000000004


observe que o resultado é `0.30000000000000004` ao invés de apenas `0.3`, isto ocorre devido a forma como os numeros de ponto flutuante são convertidos para binário, e devido ao espaço alocado na memoria, nós temos um overflow de bits

esse problema ocorre pois a parte fracionaria de um Float tem um tamanho pré-definido

quando chegamos ao fim desse tamanho e a operação não for solucionada, se existirá uma sobrecarga de bits, que por ventura culminará no numero 4 ao final da cadeia

##### Performance

os numeros de pontos flutuantes são "menos eficientes" que os numeros integros devido a forma que eles são guardados na memória

a parte inteira se localiza em um endereço, e a parte fracionária (decimal) se localiza em outra, ou seja, para performar operações o sistema deve navegar a varios endereços o que culmina num aumento do tempo de resposta, porém, vale ressaltar que este tempo só se torna perceptivel caso você tenha operações na casa das milhares, (3 milhoes de calculos, 19 milhoes de calculos, etc...) pois essa diferença de performance é na ordem dos nanosegundos

### Boolean

O Boolano é o tipo de dado Binário, só pode ter 2 valores, Verdadeiro (`True`) ou Falso (`False`)

ele é bastante usado para controlar o Flow do programa através de `if`s e `while`s

os valores boleanos são dividos em 2 categorias, Verdadeiros e Falsos

- Verdadeiros:
    - `True`
    - qualquer string que não esteja vazia
    - qualquer numero diferente de 0
    - qualquer outro tipo de dado que não esteja vazio
- Falsos:
    - `False`
    - Strings vazias ("" ou '')
    - 0
    - qualquer outro tipo de dado vazio

#### Operadores Lógicos

esses operadores são usados em conjunto dos booleanos, são 3, `and`, `or`, `not`

o `and`, retorna `True` caso todos os valores sejam verdadeiros

In [3]:
print(True and True)
print(True and False)
print(False and True)
print(False and False)

True
False
False
False


o `or` retorna `True` caso um dos valores seja Verdadeiro, ou seja, só retorna Falso caso ambos os valores sejam falsos, informalmente dizendo, podemos presumir que ele exerce o oposto do `and`

In [4]:
print(True or True)
print(True or False)
print(False or True)
print(False or False)

True
True
True
False


o `not` é o mais simples de todos, ele inverte o valor

In [6]:
print(not False)
print(not True)

True
False


#### Curto Circuito

o curto circuito em operadores lógicos é uma pratica onde você ordena os valores para que a decisão seja tomada mais rapida

por exemplo, com o `and`, se você tem algum valor que provavelmente retornará `False`, o colocar na frente o dará performance, pois o Python irá ignorar o resto da expressão pois o `and` só retorna `True` caso TODOS os valores sejam verdadeiros

com o `or`, se é exatamente o oposto, colocamos um valor que provavelmente retornará `True` na frente, pois assim a expressão já é validada, retornando `True`, e o Python por sua vez irá ignorar o resto da expressão, pois o `or` retorna `True` caso um dos valores seja verdadeiro


## Tipos Compostos

Os tipos compostos são aqueles que contem mais de um Dado

eles são os seguintes:

- Tuples
- Listas
- Sets
- Dicionarios

### Mutabilidade

A Mutabilidade define a capacidade de um tipo ser alterado sem que se seja criado um novo, exemplo:

Listas são mutaveis, pois nós podemos adicionar ou remover elementos sem criar uma lista nova

Tuples são imutaveis pois para podermos adicionar um novo elemento nós devemos criar uma nova

### Tuples

As Tuples são imutaveis, elas são comumente usadas na declaração de multiplas variaveis, para declara uma tuple nós podemos fazer de duas formas:

forma 1:
usamos parenteses `()`


In [2]:
tulipa1 = ("eba", "uma", "tuple")
print(tulipa1, type(tulipa1))

('eba', 'uma', 'tuple') <class 'tuple'>


forma 2:
nas tuples nós podemos omitir o parenteses, apenas separando os argumentos por virgula `,`

In [3]:
tulipa2 = "eba", "mais", "uma", "tuple"
print(tulipa2, type(tulipa2))

('eba', 'mais', 'uma', 'tuple') <class 'tuple'>


Para adicionarmos ou removermos elementos de uma tuple nós devemos criar uma nova

In [2]:
x, y = 1, 2 # uma tuple com 2 valores
z = 7 # valor que iremos adicionar
tulipa_nova = x, + y, + z
print(tulipa_nova, type(tulipa_nova))

(1, 2, 7) <class 'tuple'>


### Listas

Diferente das Tuples, as listas já podem ser alteradas, para declararmos uma lista nós usamos `[]`

In [5]:
amigos = ['amigo1', 'amigo2', 'amigo3']
print(amigos, type(amigos))

['amigo1', 'amigo2', 'amigo3'] <class 'list'>


para obtermos um item especifico nós usamos um index iniciado em 0

é simples, basta fazer `nome_da_lista[index]`, e as listas começam pelo 0, ou seja, o primeiro valor é 0, o segundo é 1, o terceiro é 2 e assim por diante

caso nós coloquemos um sinal negativo a frente do index, a lista será invertida, `-1` se torna o ultimo, `-2` o penultimo, `-3` o antepenultimo e assim por diante

eis aqui um exemplo:

In [10]:
amigos = ['amigo1', 'amigo2', 'amigo3']
print(amigos, type(amigos))
print(amigos[0], type(amigos[0])) # mostraremos o amigo1
print(amigos[2], type(amigos[1])) # mostraremos o amigo2
print(amigos[1], type(amigos[2])) # mostraremos o amigo3
print(amigos[-3], type(amigos[-3])) # mostraremos o amigo1
print(amigos[-2], type(amigos[-2])) # mostraremos o amigo2
print(amigos[-1], type(amigos[-1])) # mostraremos o amigo3

['amigo1', 'amigo2', 'amigo3'] <class 'list'>
amigo1 <class 'str'>
amigo3 <class 'str'>
amigo2 <class 'str'>
amigo1 <class 'str'>
amigo3 <class 'str'>
amigo2 <class 'str'>


para adicionarmos um item a uma lista nós usamos o metodo `append`, sua sintaxe é:

`lista.append(item)`, este metodo sempre adicionará o item ao final da lista

In [8]:
amigos = ['amigo1', 'amigo2', 'amigo3']
print(amigos, type(amigos))
amigos.append('amigo4')
print(amigos, type(amigos))

['amigo1', 'amigo2', 'amigo3'] <class 'list'>
['amigo1', 'amigo2', 'amigo3', 'amigo4'] <class 'list'>


para adicionarmos um item em uma posição especifica nós usamos `lista.insert(index, item)`

In [9]:
amigos = ['amigo1', 'amigo2', 'amigo3']
print(amigos, type(amigos))
amigos.insert(1, 'amigo4') # iremos adicionar o amigo4 na segunda posição
print(amigos, type(amigos))

['amigo1', 'amigo2', 'amigo3'] <class 'list'>
['amigo1', 'amigo4', 'amigo2', 'amigo3'] <class 'list'>


para removermos um item, nós temos `lista.pop()` e `lista.remove(item)`

o metodo `lista.pop()` remove o ultimo item, ou em um index especificado `lista.pop(index)`
o metodo `lista.remove(item)` remove o item especificado

In [11]:
amigos = ['amigo1', 'amigo2', 'amigo3']
print(amigos, type(amigos))
amigos.pop()
print(amigos, type(amigos))
amigos.remove('amigo1')
print(amigos, type(amigos))

['amigo1', 'amigo2', 'amigo3'] <class 'list'>
['amigo1', 'amigo2'] <class 'list'>
['amigo2'] <class 'list'>


#### Matrizes (Lista Multi-Dimensional)

Caso você se depare na situação de ter que criar uma matiz, em python isso pode ser feito através de listas multi-dimensionais

In [13]:
matriz = [
# X  0  1  2     Y
    [1, 2, 3], # 0
    [4, 5, 6], # 1
    [7, 8, 9]  # 2
]
print(matriz, type(matriz))
# Eixos:     Y  X
print(matriz[0][0])
print(matriz[2][1])

[[1, 2, 3], [4, 5, 6], [7, 8, 9]] <class 'list'>
1
8


### Sets (Conjuntos)

Sets são essencialmente conjuntos, são declarados através das chaves `{}`

as operações são, `set.add()`, `set.remove()`, `set.difference()`, `set.symmetric_difference()`, `set.intersection()`, `set.union()`

`set.add()` adiciona elementos ao set

`set.remove()` remove elementos do set

`set.difference()` retorna a diferença entre sets

`set.symmetric_difference()` retorna todos os itens que não estão na interseção

`set.intersection()` retorna a intersceção do set

`set.union()` retorna a união do set

In [3]:
estudantes_matematica = {'aluno1', 'aluno2', 'aluno2', 'aluno3', 'aluno4', 'aluno5'}

estudantes_portugues = {'aluno6', 'aluno2', 'aluno3', 'aluno7', 'aluno8'}

estudantes_matematica.add('aluno9')
print(f"ADD: {estudantes_matematica}")
estudantes_matematica.remove('aluno5')
print(f"REMOVE: {estudantes_matematica}")
print(f"DIFFERENCE: {estudantes_portugues.difference(estudantes_matematica)}")
print(f"SYMMETRIC_DIFFERENCE: {estudantes_portugues.symmetric_difference(estudantes_matematica)}")
print(f"INTERSECTION: {estudantes_matematica.intersection(estudantes_portugues)}")
print(f"UNION: {estudantes_matematica.union(estudantes_portugues)}")

ADD: {'aluno2', 'aluno9', 'aluno3', 'aluno4', 'aluno5', 'aluno1'}
REMOVE: {'aluno2', 'aluno9', 'aluno3', 'aluno4', 'aluno1'}
DIFFERENCE: {'aluno7', 'aluno6', 'aluno8'}
SYMMETRIC_DIFFERENCE: {'aluno9', 'aluno6', 'aluno8', 'aluno4', 'aluno7', 'aluno1'}
INTERSECTION: {'aluno2', 'aluno3'}
UNION: {'aluno9', 'aluno4', 'aluno1', 'aluno2', 'aluno6', 'aluno3', 'aluno8', 'aluno7'}


### Dicionários

Dicionários são listas com `chave:valor`, onde o valor pode ter um indentificador nomeado chamado chave

a declaração de um dicionário é feita da seguinte forma:



In [6]:
valor1, valor2, chave1, chave2 = "valor1", "valor2", "chave1", "chave2"

dicionario = {chave1: valor1, chave2: valor2}
print(dicionario, type(dicionario))

{'chave1': 'valor1', 'chave2': 'valor2'} <class 'dict'>


para fins de conveniencia e legibilidade, nós dividimos o dicionario entre linhas, ficando desta forma

In [7]:
valor1, valor2, chave1, chave2 = "valor1", "valor2", "chave1", "chave2"

dicionario = {
    chave1: valor1, 
    chave2: valor2
}
print(dicionario, type(dicionario))

{'chave1': 'valor1', 'chave2': 'valor2'} <class 'dict'>


para acessarmos um valor especifico nós usamos sua chave, desta forma `dicionario[chave]`

In [8]:
valor1, valor2, chave1, chave2 = "valor1", "valor2", "chave1", "chave2"

dicionario = {
    chave1: valor1, 
    chave2: valor2
}
print(dicionario[chave1], type(dicionario[chave1]))

valor1 <class 'str'>


para adicionarmos um valor, ou reassociarmos um valor, basta por `dicionario[chave]=novo_valor`

In [9]:
valor1, valor2, valor3 = "valor1", "valor2", "valor3"
chave1, chave2, chave3 = "chave1", "chave2", "chave3"

dicionario = {
    chave1: valor1, 
    chave2: valor2
}

dicionario[chave3] = valor3
dicionario[chave1] = valor2
print(dicionario[chave3], type(dicionario[chave3]))
print(dicionario[chave1], type(dicionario[chave1]))

valor3 <class 'str'>
valor2 <class 'str'>


### `sum()` & `len()`

`sum()` é uma função nativa do python que retorna a soma de todos os valores de um objeto iteravel, ou seja, listas, dicionarios, sets e tuples

`len()` é uma função nativa do python que retorna o comprimento de um objeto iteravel, ou seja, listas, dicionarios, sets e tuples

seu uso principal seria para calcular médias, pois uma média é dada através da soma de todos os valores dividos pela quantidade de valores